### DISCUSSION TITLE
SO101 仿真测评路线图：当前进展、架构与下一步

### DISCUSSION THREAD

**Discussion by xieyucheng123** [link-seek/so101-sim-pipeline#3] [General]
## 仓库定位

`so101-sim-pipeline` — SO101 LeRobot 仿真 CI/CD 流水线，目标：GitHub Actions 触发 → 华为云 V100 ECS 训练 → 仿真评估 → OBS 归档。

---

### 1.1 整体流水线

```
GitHub dispatch
  │
  ▼
start-ecs (ubuntu-latest)
  │  hcloud CLI → BatchStartServers → sleep 120s
  │
  ▼
vla-pipeline (self-hosted, Linux, X64, V100)
  │  ① SWR 登录 + 拉取镜像
  │  ② Train SmolVLA (lerobot-train, --rename_map)
  │  ③ Simulation Replay (replay_demo.py → MuJoCo)
  │  ④ Upload to OBS
  │
  ▼
stop-ecs (ubuntu-latest, if: always())
     hcloud CLI → BatchStopServers
```

### 1.2 Docker 镜像矩阵

| 镜像 | Dockerfile | 基础 | 关键依赖 | 用途 |
|------|-----------|------|----------|------|
| `so101-train` | `Dockerfile.train` | nvidia/cuda:12.6.3-runtime-ubuntu22.04 | torch 2.11 cu126 + lerobot[smolvla,training]==0.6.1 + so101_nexus[warp] + vla-eval==0.4.0 | 训练 + 回放 |
| `so101-eval` | `Dockerfile.eval` | 同上 | SQLite 3.45 + vla-eval | LIBERO/LIBERO-PRO benchmark 评估 |
| `so101-model-server` | `Dockerfile.model-server` | 同上 | 模型推理服务 | HTTP/ZMQ 推理 |

镜像构建：GHCR → SWR 双写（`docker-build.yml`），push to main 自动触发。

### 1.3 训练管线 (`train_smolvla.py`)

```
lerobot-train
  --policy.path=lerobot/smolvla_base
  --dataset.repo_id=<DATASET_REPO>
  --rename_map=<RENAME_MAP>     ← 相机名映射
  --steps=20000 --batch_size=32
  --save_freq=5000 --env_eval_freq=2000
  --wandb.enable=false
  --policy.push_to_hub=false
```

关键设计：
- `--rename_map` 解决数据集相机名与模型期望不一致问题（如 `side→camera1, up→camera2`）
- checkpoint 保存到 `/data/checkpoints/smolvla/checkpoints/`
- 支持 HF Hub 推送（可开关）

### 1.4 仿真回放管线 (`replay_demo.py`)

```
so101_nexus MuJoCo 环境
  ↓ obs = {state, wrist_camera, overhead_camera}
  ↓ sim_qpos_to_dataset_row()   ← 仿真角度 → 数据集单位
  ↓ prepare_observation_for_inference()
  ↓ preprocess → policy.select_action → postprocess
  ↓ dataset_row_to_sim_qpos()   ← 数据集单位 → 仿真角度
  ↓ env.step(action_rad)
  ↓ env.render() → video frames
```

环境配置：
- `MuJoCoPickAndPlace-v1`，`PickAndPlaceConfig`
- 观测：`JointPositions` + `WristCamera(640×480)` + `OverheadCamera(640×480)`
- 控制模式：`pd_joint_pos`
- 渲染：`MUJOCO_GL=egl`（无头 GPU 渲染）
- 输出：`replay_pickplace.mp4` + `replay_pickplace.json`

### 1.5 VLA 评估管线 (`eval_vla.py`)

```
start model server (vla-eval serve)
  ↓
for benchmark in [libero_spatial, libero_object, libero_goal,
                   libero_pro_swap, libero_pro_object, ...]:
    vla-eval run --config <benchmark>.yaml
    vla-eval merge
  ↓
stop model server
  ↓
generate_summary() → eval_summary.json
```

覆盖 LIBERO (3 benchmark) + LIBERO-PRO (5 benchmark)，通过 `vla-eval` harness 统一调度。

---

### 2.1 流水线状态

| Pipeline | 状态 | Run 数 | 最近结果 |
|----------|------|--------|----------|
| Docker Build | ✅ 稳定 | 2 | main 分支成功 |
| CI | ✅ 稳定 | 6 | 全部成功 |
| VLA Pipeline | ⚠️ 修复后稳定 | 20 | main 5 连败 → smolvla-fresh 2 连胜 |

VLA Pipeline 失败原因（main 分支）：
1. `--eval_freq` 参数名错误 → 修复为 `--env_eval_freq`（lerobot 0.6.1 API）
2. 训练前目录未清理 → `rm -rf /data/checkpoints/smolvla`
3. `push_to_hub` 失败（HF 不可达） → 禁用

#### 实验 1：100 steps 快速验证（[run 31675494848](https://github.com/link-seek/so101-sim-pipeline/actions/runs/31675494848)）

| 指标 | 旧方案 (side+up, 20k) | 新方案 (wrist+overhead, 100) |
|------|----------------------|----------------------------|
| Prediction errors | >0 | **0/300** |
| Reward | ≈0 | ≈0 |
| Success | False | False |

**结论**：相机匹配后模型能全程控制机器人（0 errors），P0 确认。

#### 实验 2：5k steps 训练验证（[run 31681100500](https://github.com/link-seek/so101-sim-pipeline/actions/runs/31681100500)）

训练 Loss 持续收敛：

| Step | Loss |
|------|------|
| 500 | 0.282 |
| 1000 | 0.200 |
| 2000 | 0.142 |
| 3000 | 0.105 |
| 5000 | 0.084 |

回放结果：
- Prediction errors: **0/300**（模型全程控制）
- Reward: ±0.003（仍接近 0）
- Success: **False**
- First action (ds): `[-1.24, -81.72, 84.37, 41.40, 2.58, 1.06]`（合理关节角度）

### 2.3 问题诊断

| 问题 | 严重度 | 状态 | 说明 |
|------|--------|------|------|
| P0: 相机视角不匹配 | 🔴 | ✅ 已修复 | 训练 side+up / 推理 wrist+overhead → 切换数据集 |
| P1: Sim-to-Real Visual Gap | 🔴 | ❌ 未解决 | 真机照片训练 → MuJoCo 渲染评估，视觉域不匹配 |
| P2: 训练步数不足 | 🟡 | ❌ 未解决 | 5k 步 loss 仍在下降，SmolVLA 标准需 20k-30k |
| P3: 数据采集方式 | 🟡 | ⚠️ | 硬编码 keyframe 轨迹，质量不如遥操作 |

---

## 三、开源生态对比

| 方案 | 训练数据 | 评估环境 | 视觉域 | 代表项目 |
|------|----------|----------|--------|----------|
| **我们当前** | 真机照片 → 仿真评估 | MuJoCo | ❌ 不匹配 | so101-sim-pipeline |
| LeRobot/SmolVLA 官方 | 真机数据 → 真机评估 | 真机 | ✅ | lerobot |
| LIBERO/ManiSkill | 仿真数据 → 仿真评估 | 仿真 | ✅ | libero, maniskill3 |
| SO-101 Bench | 真机+仿真 → 仿真数字孪生 | Isaac Lab | ✅ | 5hadytru/so101_bench |
| so101_nexus 官方 | MuJoCo 遥操作 → MuJoCo 评估 | MuJoCo | ✅ | johnsutor/so101_nexus |
| NVIDIA Workshop | 仿真遥操作 → Isaac Sim 评估 | Isaac Sim | ✅ | isaac-sim/Sim-to-Real-SO-101-Workshop |

**核心原则**：训练和评估必须在同一视觉域。我们当前的真机→仿真跨域是 success=0% 的根因。

---

### 优先级排序

| 优先级 | 方案 | 解决问题 | 预期耗时 | 风险 | 依赖 |
|--------|------|----------|----------|------|------|
| **P0** | 切换 MuJoCo 仿真数据集 | P1 Visual Gap | 采集 0 + 训练 3.5h (5k) | 10 episodes 偏少 | Discussion #1 评论方案 |
| **P1** | 5k→20k 完整训练 | P2 训练不足 | +10h | P1 仍存在若未切数据集 | 先完成 P0 |
| **P2** | PPO 纯仿真 baseline | 环境可学性验证 | ~50min (V100) | MLP 策略，非 VLA | Discussion #2 方案 |
| **P3** | 仿真数据补充采集 | P3 数据质量 | 采集 30min | 硬编码轨迹质量 | so101_nexus 遥操作 |

### P0：切换 MuJoCo 仿真数据集（推荐立即执行）

**目标**：消除 P1 Visual Gap

**数据集**：`johnsutor/MuJoCoPickAndPlace-v1`
- 来源：so101_nexus 官方 MuJoCo 仿真遥操作
- 相机：wrist + overhead（与回放环境完全匹配）
- 格式：LeRobot v3.0，SO-101，6 DOF
- Episodes：10（偏少，NVIDIA 建议最低 75）

**执行步骤**：
1. 修改 `vla-pipeline.yml`：`DATASET_REPO=johnsutor/MuJoCoPickAndPlace-v1`，`RENAME_MAP` 更新
2. 触发 5k steps 训练验证
3. 对比 reward/success 与当前方案

**风险**：10 episodes 可能不够 SmolVLA 学习，需观察 loss 曲线和回放 reward

### P1：20k 完整训练

在 P0 验证数据集可行后，跑 20k-30k steps 完整训练。V100 预估 ~14h。

### P2：PPO 纯仿真 baseline

**目标**：验证 WarpPickLift-v1 环境可学性，建立 RL baseline

**交付物**（详见 Discussion #2）：
1. `docker/Dockerfile.ppo`
2. `scripts/train_ppo.py`（从 so101_nexus `ppo_warp.py` 适配）
3. `scripts/eval_ppo.py`（MuJoCo 50 episodes 评估）
4. `.github/workflows/ppo-pipeline.yml`
5. `docker-build.yml` 追加 build-ppo job

**官方参考结果**：RTX 5090 上 WarpPickLift-v1 success 0.973 (5 seed)，V100 预估 ~50min。

### P3：仿真数据补充采集

如果 P0 的 10 episodes 不够：
- 短期：`collect_sim_data.py` 硬编码轨迹补充（质量差但量大）
- 长期：so101_nexus 遥操作采集（需物理 leader arm，我们没有）
- 替代：寻找更多 MuJoCo 仿真遥操作公开数据集

---

## 五、决策点

1. **P0 是否立即执行？** — 切换到 `johnsutor/MuJoCoPickAndPlace-v1`，5k 验证
2. **PPO baseline 是否并行启动？** — 与 VLA 路线独立，可同时推进
3. **最终目标是真机部署还是纯仿真验证？** — 决定后续是否需要 sim2real 迁移
4. **10 episodes 不够时怎么办？** — 补充采集方案选择

---

## 六、时间线预估

```
Week 1:  P0 执行（MuJoCo 数据集切换 + 5k 验证）
         P2 启动（PPO 流水线代码 + docker build）
Week 2:  P1 执行（20k 完整训练）
         P2 验证（PPO pipeline 首次 run）
Week 3:  根据结果决定 P3（数据补充）或转向真机部署
```

---

*关联 Discussion：[#1 SmolVLA 仿真回放验证](https://github.com/link-seek/so101-sim-pipeline/discussions/1) | [#2 PPO 纯仿真流水线设计](https://github.com/link-seek/so101-sim-pipeline/discussions/2)*

**Comment by xieyucheng123** [comment]
## 收尾（已被取代）

本路线图描述的是老架构（单 `vla-pipeline` 线）。架构已演进为 `franka-eval` / `robot-switch` / `collect` / `finetune` / `verify` 五条线，计划跟踪已移到 #16。

本帖 OUTDATED 关闭，后续看 #16。
